### PR TITLE
Exponential back-off for process restarting

### LINKED ISSUE #265: Feature Request: Exponential backoff
Would be nice to have an exponential backoff feature when a process fails and you would like to restart it, but then allow that exponential backoff to be reset when the process is "alive" for a certain amount of time. As a start, this could even be simply allowing restartDelay option to take a function instead of a number.

Example:
- process fails -> immediate restart
- process fails -> restart after 2 seconds
- process fails -> restart after 4 seconds
- process fails -> restart after 8 seconds
- process "succeeds", where "success" is defined as "did not exit for 60 seconds after starting"
- process fails -> immediate restart
- process fails -> restart after 2 seconds

### PR DESCRIPTION
**Pull request by gustavohenke**
Title.

```
$ conc --restart-tries 4 --restart-after exponential "date && exit 1"
[0] Fri 12 Jan 2024 14:10:56 -03
[0] date && exit 1 exited with code 1
[0] date && exit 1 restarted
[0] Fri 12 Jan 2024 14:10:57 -03
[0] date && exit 1 exited with code 1
[0] date && exit 1 restarted
[0] Fri 12 Jan 2024 14:10:59 -03
[0] date && exit 1 exited with code 1
[0] date && exit 1 restarted
[0] Fri 12 Jan 2024 14:11:03 -03
[0] date && exit 1 exited with code 1
[0] date && exit 1 restarted
[0] Fri 12 Jan 2024 14:11:11 -03
[0] date && exit 1 exited with code 1
```

Closes #265

### CODE REVIEW

File: src/flow-control/restart-process.spec.ts [hunk 1]
```diff
@@ -8,12 +8,14 @@ import { RestartProcess } from './restart-process';
...
+let scheduler: VirtualTimeScheduler;
 beforeEach(() => {
     commands = [new FakeCommand(), new FakeCommand()];
-
     logger = createMockInstance(Logger);
-    scheduler = new TestScheduler(() => true);
+
+    // Don't use TestScheduler as it's hardcoded to a max number of "frames" (time),
+    // which don't work for some tests in this suite
+    scheduler = new VirtualTimeScheduler();
```
Comment (gustavohenke) [inline src/flow-control/restart-process.spec.ts:18]: Evidence https://github.com/ReactiveX/rxjs/blob/7.8.1/src/internal/testing/TestScheduler.ts#L18

File: src/flow-control/restart-process.ts [hunk 2]
```diff
@@ -21,13 +23,13 @@ export class RestartProcess implements FlowController {
...
     }: {
-        delay?: number;
+        delay?: RestartDelay;
         tries?: number;
         logger: Logger;
         scheduler?: Rx.SchedulerLike;
     }) {
         this.logger = logger;
-        this.delay = delay != null ? +delay : defaults.restartDelay;
+        this.delay = delay != null ? delay : 0;
```
Comment (gustavohenke) [inline src/flow-control/restart-process.ts:32]:
```suggestion
        this.delay = delay ?? 0;
```

**Comment by mscottnelson** [comment]
Very nice! Looks great! Thank you for adding this feature

**Comment by gustavohenke** [comment]
🚢 This is now available in v9.0.0!
https://github.com/open-cli-tools/concurrently/releases/tag/v9.0.0

**Pull request merged**